Refactor & Support code block with `<C-e>`

ful1e5 opened the pull request:
This refactor involves several modifications to the codebase aimed at enhancing **code quality and readability**. Key changes include the addition of Lua documentation to functions, the renaming of internal variables and functions, and the modification of the way position variables are accessed. The `surrounder` function has also been **heavily refactored** and renamed to `__toggle_surround` to reduce the number of lines of code and improve readability.

The codebase also includes a **new feature** that supports formatting **code blocks** using the `<C-e>` key combination (fixes #4). The feature consists of an implementation of a code function for handling **single-line** and **multi-line** delimiters using a table-based approach, which allows for more flexible handling of parameters. Additionally, the `__update_start_selection_mark` utility function has been introduced to facilitate updating the starting position of the selection mark when using optional parameters such as adding surround to multiple lines of code.


Further details on these changes can be found in individual commits.

https://user-images.githubusercontent.com/24286590/215761705-97d2156e-877d-4413-8e38-05b3bc78d0a8.mp4

ful1e5 committed:
refactor: added lua docs and renamed internal variables & functions

THIS REFACTOR DOES NOT MAKE ANY CHANGES TO LOGIC OR FEATURES

This refactor addresses several key areas of the codebase. The following
changes have been made:

- Lua documentation has been added to functions and access parameters
  have been assigned to them.
- The function return has been removed from `make_surrounder_function`
  and it has been renamed to `__wrap_sel_text`
- Two new functions, `__buf_get_single_line` and `__buf_set_single_line`
  have been added to perform read and write operations on a single line.
- The way position variables are accessed has been refactored by using
  tables (e.g. `pos_start = { line=1, col=2 }`) to enhance code
  readability.
- The `surrounder` function has been heavily refactored by renaming it
  to `__toggle_surround` and moving some variables to the function scope
  to reduce the number of lines of code and improve readability.
- Several variable names (e.g. `post_start` to `start_pos`) and some
  function names have been renamed for improved readability and
  consistency.

ful1e5 committed:
feat: support code block using <C-e>

This patch addresses issue #4 and includes additional codebase
refactoring, including the following:

- Implementation of a code function for formatting code blocks with
  either single-line ('`') or multi-line ('```') delimiters.
- Conversion to a table-based approach for delimiter management,
  allowing for more flexible handling of parameters such as multiline
  surrounds `mln_sur` and `opts` optional parameters.
- Introduction of the `__update_start_selection_mark` utility function
  to facilitate updating the starting position of the selection mark
  ('<'), which is particularly useful when using the optional newline
  parameter for adding surround to multiple lines of code.

antonk52 commented:
Nice work, the only change I want to ask is to make is when an entire line is selected to add the triple backtick to new lines surrounding the selected line instead of on the same line. Personally I haven't came across such usage in my career. Though if you think it is a useful case, maybe we should have an option to enable `single_line_code_block`. What do you think?

ful1e5 committed:
Merge branch 'main' into feat-code-block

ful1e5 committed:
breaking-change: Entire plugin rewritten and `inline_code()` deprecated

I (@ful1e5) have completely overhauled the plugin to accommodate newline
surroundings. Much of the code was drawn from the "nvim-surround.buffer"

As a result of this change, the `inline_code()` function has been
deprecated and should no longer be used. Instead, use the `code()`
function.

ful1e5 committed:
breaking-change: Keybind for toggling code block changed to `<C-e>`

The keybinding for toggling a code block has been changed to `<C-e>` due
to the fact that many Vim users have remapped `<C-c>` to the `Esc` key.
Additionally, `<C-e>` is the default keybinding for code blocks on
platforms such as GitHub.

ful1e5 commented:
I completely rewrote the plugin and deprecated the `inline_code()` function. I have added newline and single-line surroundings like you asked, considering #1.

If you want to disregard older commits, I am happy to open a new Pull Request with only the latest commits.

ful1e5 commented:
Closing PR; because #7.
